yente/search/base.py

```diff
@@ -74,3 +91,317 @@ async def close_es() -> None:
+        """
+        Return a document delete payload that can be accepted by the concrete search provider.
+        """
+        pass
+
+    @abstractmethod
+    def _create_operation(
+        self, entity: Dict[str, Any], doc_id: str, index: str
+    ) -> Dict[str, Any]:
+        """
+        Return a document create payload that can be accepted by the concrete search provider.
+        """
+        pass
+
+    @abstractmethod
+    def _update_operation(
+        self, entity: Dict[str, Any], doc_id: str, index: str
+    ) -> Dict[str, Any]:
+        """
+        Return a document update payload that can be accepted by the concrete search provider.
+        """
+        pass
+
+    @abstractmethod
+    async def get_backing_indexes(self, name: str) -> List[str]:
+        """
+        Get all the indexes backing Yente search. In ElasticSearch this would be implemented
+        with multiple indexes pointing to the alias specified by the name parameter.
+        """
+        pass
+
+    def _to_operation(self, body: Dict[str, Any], index: str) -> Dict[str, Any]:
+        """
+        Convert an entity to a bulk operation.
+        """
+        try:
+            entity = body.pop("entity")
+            doc_id = entity.pop("id")
+        except KeyError:
+            raise ValueError("No entity or ID in body.\n", body)
+        match body.get("op"):
+            case "ADD":
+                return self._create_operation(entity, doc_id, index)
+            case "MOD":
+                return self._update_operation(entity, doc_id, index)
+            case "DEL":
+                return self._delete_operation(doc_id, index)
+            case _:
+                raise ValueError(f"Unknown operation type: {body.get('op')}")
+
+
+class ESSearchProvider(SearchProvider):
+    @classmethod
+    async def create(cls) -> "ESSearchProvider":
+        self = cls()
+        self.client = await get_es()
+        return self
+
+    async def upsert_index(self, index: str) -> None:
+        """
+        Create an index if it does not exist. If it does exist, do nothing.
+        """
+        try:
+            schemata = list(model.schemata.values())
+            mapping = make_entity_mapping(schemata)
+            await self.client.indices.create(
+                index=index, mappings=mapping, settings=INDEX_SETTINGS
+            )
+        except BadRequestError:
+            pass
+
+    async def _remove_write_block(self, index: str) -> None:
+        await self.client.indices.put_settings(
+            index=index, settings={"index.blocks.read_only": False}
+        )
+
+    async def _add_write_block(self, index: str) -> None:
+        await self.client.indices.put_settings(
+            index=index, settings={"index.blocks.read_only": True}
+        )
+
+    async def clone_index(self, index: str, new_index: str) -> None:
+        try:
+            await self._add_write_block(index)
```

Comment:
I wonder if the write block should be added after the index is written, not before it is cloned.

Reply:
ES won't let you clone the index if it isn't read only, but it's a good idea to make 100% sure there is no missed writes on the clone by doing the rollover before we remove the write block. So it's bee 
1. Add write block
2. Clone index
3. Update clone
4. Rollover to clone
5. Delete the old index
If there is an exception, delete the clone and remove the write block from the original index.